rows_test.go

```diff
@@ -0,0 +1,319 @@
+			want: []testAtomicTypesRow{
+				{key: "pinf", testString: "hello", testBytes: []byte("hello"), testInt: 42, testFloat: math.Inf(1), testBool: true},
+			},
+		},
+		{
+			name:  "special float negative infinity",
+			input: `SELECT * FROM TestAtomicTypes WHERE key = "ninf"`,
+			want: []testAtomicTypesRow{
+				{key: "ninf", testString: "hello", testBytes: []byte("hello"), testInt: 42, testFloat: math.Inf(-1), testBool: true},
+			},
+		},
+	}
+
+	// Run tests.
+	for _, tc := range tests {
+
+		rows, err := db.QueryContext(ctx, tc.input)
+		if (err != nil) && (!tc.wantError.query) {
+			t.Errorf("%s: unexpected query error: %v", tc.name, err)
+		}
+		if (err == nil) && (tc.wantError.query) {
+			t.Errorf("%s: expected query error but error was %v", tc.name, err)
+		}
+
+		got := []testAtomicTypesRow{}
+		for rows.Next() {
+			var curr testAtomicTypesRow
+			err := rows.Scan(
+				&curr.key, &curr.testString, &curr.testBytes, &curr.testInt, &curr.testFloat, &curr.testBool)
+			if (err != nil) && (!tc.wantError.scan) {
+				t.Errorf("%s: unexpected scan error: %v", tc.name, err)
+			}
+			if (err == nil) && (tc.wantError.scan) {
+				t.Errorf("%s: expected scan error but error was %v", tc.name, err)
+			}
+
+			got = append(got, curr)
+		}
+
+		rows.Close()
+		err = rows.Err()
+		if (err != nil) && (!tc.wantError.close) {
+			t.Errorf("%s: unexpected rows.Err error: %v", tc.name, err)
+		}
+		if (err == nil) && (tc.wantError.close) {
+			t.Errorf("%s: expected rows.Err error but error was %v", tc.name, err)
+		}
+
+		if !reflect.DeepEqual(tc.want, got) {
+			t.Errorf("Unexpected rows: %s. want: %v, got: %v", tc.name, tc.want, got)
+		}
+
+	}
+
+	// Drop table.
+	_, err = db.ExecContext(ctx, `DROP TABLE TestAtomicTypes`)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+}
+
+func TestRowsOverflowRead(t *testing.T) {
+
+	// Open db.
+	ctx := context.Background()
+	db, err := sql.Open("spanner", dsn)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	// Set up test table.
+	_, err = db.ExecContext(ctx,
+		`CREATE TABLE TestOverflowRead (
+			key	STRING(1024),
+			testString	STRING(1024),
+			testBytes	BYTES(1024),
+			testInt	INT64,
+			testFloat	FLOAT64,
+			testBool	BOOL
+		) PRIMARY KEY (key)`)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = db.ExecContext(ctx,
+		`INSERT INTO TestOverflowRead  (key, testString, testBytes, testInt, testFloat, testBool) VALUES
+		("general", "hello", CAST ("hello" as bytes), 42, 42, TRUE), 
+		("maxint", "hello", CAST ("hello" as bytes), 9223372036854775807 , 42, TRUE),
+		("minint", "hello", CAST ("hello" as bytes), -9223372036854775808 , 42, TRUE),
+		("max int8", "hello", CAST ("hello" as bytes), 127 , 42, TRUE),
+		("max uint8", "hello", CAST ("hello" as bytes), 255 , 42, TRUE)
+		`)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	type testAtomicTypesRowSmallInt struct {
+		key        string
+		testString string
+		testBytes  []byte
+		testInt    int8
```

Comment:
Wondering if overflow / underflow is with the same behaviour as other drivers?

Reply:
Yeah I was wondering that too, I'll confirm with pg. It's slightly odd, scans until it gets to an err then stops, returns the half scanned row

Reply:
pg does the same thing 👍